lib/util/source-code.js

```diff
@@ -191,25 +202,61 @@ SourceCode.prototype = {
      * @public
      */
     getComments(node) {
+        if (this._commentStore.has(node)) {
+            return this._commentStore.get(node);
+        }
 
-        let leadingComments = node.leadingComments || [];
-        const trailingComments = node.trailingComments || [];
+        const comments = {
+            leading: [],
+            trailing: []
+        };
 
         /*
-         * espree adds a "comments" array on Program nodes rather than
-         * leadingComments/trailingComments. Comments are only left in the
-         * Program node comments array if there is no executable code.
+         * Espree only leaves comments in the Program node comments
+         * array if there is no executable code.
          */
         if (node.type === "Program") {
             if (node.body.length === 0) {
-                leadingComments = node.comments;
+                comments.leading = node.comments;
+            }
+        } else if ((node.type === "BlockStatement" || node.type === "ClassBody") && node.body.length === 0 ||
+            node.type === "ObjectExpression" && node.properties.length === 0 ||
+            node.type === "ArrayExpression" && node.elements.length === 0 ||
+            node.type === "SwitchStatement" && node.cases.length === 0
+        ) {
+            const tokens = this._getTokens(node);
+
+            comments.trailing = tokens.filter(astUtils.isComment);
```

Comment:
Worked on this today and I'm actually not sure we want to do this. There are some rules - such as `lines-around-comments` - that rely on visiting comment nodes, and ignoring these cases means that the comments in the following cases are not evaluated (and the tests fail): https://github.com/eslint/eslint/blob/master/tests/lib/rules/lines-around-comment.js#L921-L938.

There are a few other rules with failing tests (no-empty and no-empty function), however these rules can be fixed by using token and comment iterators. I'm not sure what we do for the above case - the benefit of the way Espree handled this is that these comments didn't get lost, even if it led to some unpredictable behavior.

I'm a little bit at a loss right now as to what the best way forward is. I think the behavior we discussed above for `sourceCode.getComments()` makes a lot of sense - the issue is that we also use that method to find comment tokens to visit. I wonder if this could be fixed by breaking the tie between `CommentEventGenerator` and `sourceCode.getComments()`?

I'm not sure of the the performance implications, but one thought I had was that we could iterate over all the tokens and comments of a given node in the `CommentEventGenerator` method, collect the comment tokens, and emit events for those comments. We'd still have to figure out a way to avoid emitting duplicate events as the traversal occurs, though.